Code/Framework/AzCore/AzCore/DOM/DomPath.cpp

```
@@ -0,0 +1,422 @@
+    }
+
+    const Path::ContainerType& Path::GetEntries() const
+    {
+        return m_entries;
+    }
+
+    void Path::Push(PathEntry entry)
+    {
+        m_entries.push_back(AZStd::move(entry));
+    }
+
+    void Path::Push(size_t entry)
+    {
+        Push(PathEntry(entry));
+    }
+
+    void Path::Push(AZ::Name entry)
+    {
+        Push(PathEntry(AZStd::move(entry)));
+    }
+
+    void Path::Push(AZStd::string_view entry)
+    {
+        Push(AZ::Name(entry));
+    }
+
+    void Path::Pop()
+    {
+        m_entries.pop_back();
+    }
+
+    void Path::Clear()
+    {
+        m_entries.clear();
+    }
+
+    PathEntry Path::At(size_t index) const
+    {
+        if (index < m_entries.size())
+        {
+            return m_entries[index];
+        }
+        return {};
+    }
+
+    size_t Path::Size() const
+    {
+        return m_entries.size();
+    }
+
+    PathEntry& Path::operator[](size_t index)
+    {
+        return m_entries[index];
+    }
+
+    const PathEntry& Path::operator[](size_t index) const
+    {
+        return m_entries[index];
+    }
+
+    Path::ContainerType::iterator Path::begin()
+    {
+        return m_entries.begin();
+    }
+
+    Path::ContainerType::iterator Path::end()
+    {
+        return m_entries.end();
+    }
+
+    Path::ContainerType::const_iterator Path::begin() const
+    {
+        return m_entries.cbegin();
+    }
+
+    Path::ContainerType::const_iterator Path::end() const
+    {
+        return m_entries.cend();
+    }
+
+    Path::ContainerType::const_iterator Path::cbegin() const
+    {
+        return m_entries.cbegin();
+    }
+
+    Path::ContainerType::const_iterator Path::cend() const
+    {
+        return m_entries.cend();
+    }
+
+    size_t Path::GetStringLength() const
+    {
+        size_t size = 0;
+        for (const PathEntry& entry : m_entries)
+        {
+            ++size;
+            if (entry.IsEndOfArray())
+            {
+                size += 1;
```

There shouldn't be a pathEntry after the end of array right? In that case, can we throw an error here if that isn't the case?

---

Lookup on paths with end of array markers might be invalid, but atm I'm not enforcing any constraints on the makeup of the path itself - good shout though, we might want to add more safety around string conversion later if we end up serializing these a lot.